src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1218,21 +1211,21 @@ private static unsafe string NegativeInt32ToDecStr(int value, int digits, string
 
             int bufferLength = Math.Max(digits, FormattingHelpers.CountDigits((uint)(-value))) + sNegative.Length;
             string result = string.FastAllocateString(bufferLength);
-            fixed (char* buffer = result)
-            {
-                char* p = UInt32ToDecChars(buffer + bufferLength, (uint)(-value), digits);
-                Debug.Assert(p == buffer + sNegative.Length);
+            Span<char> buffer = new Span<char>(ref result.GetRawStringData(), bufferLength);
 
-                for (int i = sNegative.Length - 1; i >= 0; i--)
-                {
-                    *(--p) = sNegative[i];
-                }
-                Debug.Assert(p == buffer);
+            int p = UInt32ToDecChars(buffer.Slice(0, bufferLength), (uint)(-value), digits);
+            Debug.Assert(p == sNegative.Length);
+
+            for (int i = sNegative.Length - 1; i >= 0; i--)
```

Comment:
Reverse iterated loops don't have bounds check elimination in the JIT today. This means we are now doing two bounds checks per loop iteration here.

It may be better to make this a normal `i = 0; i < sNegative.Length; i++` loop. Likewise, it may be good to have some check that allows the JIT to track that `buffer.Length >= sNegative.Length` so both checks are eliminated.

Reply:
Do you know why we need to calculate p, if it should be always equal to sNegative.Length?

Reply:
Done (in 4 places). Comparison: [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKBuIGYACMxgYUYG8bGfeeAHKAEsAbtgwxm5JFIAMjAHIwA5uNEwAkgDsM9UgBUIAERhgAyhigAKITsZiANgFcYaRrYyMAJkOVCMuG7E5PK4SqoY6gCU3Hy8XNRxSe4AZoxWPn4BjAA8jOQxicnFmf64jAC8+QDcNLHFPB6MwE4pKTBQADIwWsoYABaVjACy4v0AdKMIGb5lbh7jAArYULgwVgBEG1FRjADUjGEqaiIw4929A7VFDXKMsLhODp5VWjAA7nJWAOSM324tNodC59fpRa63HhmfjYLQ5MD9FYAPmarXaUCGb0+0Nh8MRUCRVgeTww40MrHxAEEoFBsABPKw7a71BpNfhDACq2l0BmMpgpK1wVkB6PGZgcQjA61kALRwJ6oKibisTg8USscEcLiV3lmAXBdRuDRS0HSTSEQyOEXU5wVA0YcBq7kYSKqsmq7jgcEKkM4LN9Io6AG0vfwALqW8InGBBoRhiGQgC+ht9xAA7PcYI9ngmksmjf7eIJROJJMEZMF5FHIqduXpDCZzJZSDY7FrXO47KUAkEQodq9FCzwEr6hGkZllynkCkOGt3ylVyMyjcUmoGunbBlVRgNJthpvP5joloL1lsdvt+8ca2cQVdZ3xK5nsy9GFivr9/qigRvLmDcw0OJwgiyLfuimIfIwQF4siRJZiSZIQAKUDUrSDJMimkJspyda8o2yFCuuYoSlKVgymB8p/jqKpqhq7Y6vOBq0CuyQmhiraeBaboehaeRWtGtp/owABkwnOnkRF3v0PF7HsPqQiOvq8OusYRlU/E3qpAHFPmD68Omz4ktpvD5nws7FmIEhSDIwAQBADiMPoUB0gOtY6PWfJNtYTTtkeniHncGnqG40EgQS3hZpEWhqBAWhuBATicXYYW4AA6sIGASFo8lJIptxjuk865PkOVKUVi7LqOdjrlJQw7hMUwTnMnakssqxntsuwHEFpyCaCxlJAVwpyr+oIuhFuBRTFWh9QMpXFHlSkpel/hZUM7p6UkBkpNgDhrANcS6SxSTLRla1VDVm6VVhdjslUXLuXh/L4kKXiRbY03EZK0qyj+UnUaqOjqpqu3am4jHXbcbFmnYXFXtavW1Y6S7Oq6jDup63qbX6x1zu90WRLFIZwOGkbXuoWnY0dqYZpYLgHaZvDmcIllltIzR2Q5TkueTbk8g2pgWFALY+aDHZNAFT49R2oX4iib2TR9hNxYwCVJZ4p2rT083xNjQ1FdOOu3OVNSYfl1UjbV25jHuB56oELUnu1mydZe0uzdJevjpdQny/jn3/dji2+prmU9OtB3FNtu37VTZsNKH50UaN97MVVnh3YwD3855BEZP7ytfaR5E+4qyqAxgwP0eD9tMdj0Mcc63HifDAm1aJLcK1NhdSTJclB9jfBd0rQhE3GZMIzGcaRyZ8dR7TUD07OjM8MzJZWeWdyuZoOioALXlWA4sXKJ2/CJX5uqTr2oTb0bOOjuOBslQPuMlPbQxLnHaeQk8x/2GLQwVRH16MDWwZ9K6Q1ZBbP6m46o20agFBYbU1guwvN1beHsZ48CfMSZ4EFPiVh+H8X66J/pYKgjCYCctk74IobiMKhJcGknJFSGk9JGR11frwbCF1LZXWxu8foQgHCSCsAACV8P0PQVh6KMAAIRujvnwYOkJM7Zw8vhF6w0fxFx+owfgOpuSoCMKIYYEAvDkAAKIAE44JpGrowaxTElK8CKnAKo1jyErwaGo3C+885EXFN9MibgDFuE6BAd40iHEQznqxU0jc4buyRk6C0aMMZCC9Eo3WXCkgqVDGpVumlp5f2xgZJhDMWRr1ZtZLevMd4YD3p5IWLZgEnzAefFql9mpS1vrOFRg1H7v0Ni/FxPATaf1yYwamP82n/2cJIKoQDj6gK0OAzh6dk5WxGPA/cTUeyO2QR1NBRSbS9zKX2JhtDCGfhIZRRU5DZagXXLQp5BI4IvkQshVC7CMLf3NhnIYpdU6+kEcI0REjlBSJbLIhR6NsnDkHkWHCj1/FaMCSRPRYTGBGJQCYkQZiLE2LsfMsGjjnFjLcR4rx2NfGotzuikauiQn6J1BEqJMKxYMVrpA4oDdzQRxbsk2BHdeLJ0we4WSCL75jJUuPdS29KZTJmbccp8EczLyqbjCypZam2Xso5Zy29cX7yFofP+HSMAX0ln2aWIVKEwXCsPAmo8VZqy6YnbW/SvaFWGc/KZAy5zvwqqUqZv9eiksWekcNyhVnrN5ckNcfChLW13Ag+2fknYoPPF1U5iN+FTKGsCwYftFYutih7aVgbiieq0BHJF2CMw7T2jAGl/yE4vRWmHOtvCYF/gTYNW6QLk39QEUIkR6RIXQpkVy+RijRljLpTnTRgp85ls+kE4uoTDG7zxaY8xVjbGwHsVytwTjyFD3fu4xxbbfRLo0c9VdzqN2YpZdi9l0TT3dP1AOuI/LYYTzbrA5GPEXTN0yVjANDaJrd1dcTUmCr6lKt9CqhoBk6ats1QWbVLNdWb31VzI19STXNObOaiNlrrUZsCtve19DqHPuVvFRKHrO1nS9bjat3ChmTmKjOKDUzL28ZDcquJSQY2RsATGuNiUNk3U8MWuBaa9mIOPEc1BubhX9p9do0hsDS2wYrYHATYzGC1vrYJ/STaY6YdE+2mtbGtY9q2QWvWQ7e16e01MsFE7xGSM/Qs3YcLZBVug/ep6YA86Mddcy8i2LcX4sJYeklDjz3QapTer+d6UXLsfasNdhmZqbqxWyyJAWyWxPs/E9iArwPFS02NUVxVotGaupK/uJnKUFzg/KvNU94yhppowDDlTsNmSNAAegmyYfgPQ3paDAEILMIAtXFB1RvdmFdGAfpbBJ2RFQUQ0SBu2X9622YyC21O6Re3Z0HejbRGdCyRLowQAAMXex997shvs/e+7sJEKI9ByaSGd2pW2Ev7qJUemAaQbuBe9WGposB8BSdou2RgABSfIv2ccUtuOjibi4ce/fIWq/Ao3xurh0B0aKDlN5NHURFvObyUTrjcFt3yXTGII82euSxWgvDaDeggYdOjznefHaIjUcAirpMYIQQgkngu7A7hNgAVIwAA8loBwdJhsQEYDACUb0OZOAF+UBEpgADWjA1cTfu0DaXfOBdC5gAgXYeQjuV0kpuULlnGh2GR9gWwb0oAXt4I97UqOgb1XGPigASjAfAVgOJRE5yEZjngg8h46HjpSKlneC4F27wpVgwrqmz8XjEBxviyG+HnnSZSMyF9dwgUbiYgA=)

I also added  "&& i < buffer.Length", what made code smaller